File: Tests/GlasstualTests/TVCMemberListSectionTests.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2008 - 2010 Satoshi Nakagawa <psychs AT limechat DOT net>
 * Copyright (c) 2010 - 2018 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import <XCTest/XCTest.h>

#import "GLTTestClient.h"
#import "IRCChannelMemberListControllerPrivate.h"
#import "IRCChannelUserPrivate.h"
#import "IRCUser.h"
#import "TVCMemberListPrivate.h"

NS_ASSUME_NONNULL_BEGIN

/* Row geometry of the member list: header rows are interleaved with
 member rows only while more than one rank is present, and every
 insertion and removal keeps the two in step. */
@interface TVCMemberListSectionTests : XCTestCase
@property(nonatomic, strong) GLTTestClient *client;
@property(nonatomic, strong) TVCMemberList *memberList;
@property(nonatomic, strong) IRCChannelMemberListController *controller;
@end

@implementation TVCMemberListSectionTests

- (void)setUp
{
	[super setUp];

	self.client = [GLTTestClient testClient];

	self.memberList = [[TVCMemberList alloc] initWithFrame:NSMakeRect(0.0, 0.0, 150.0, 400.0)];

	[self.memberList addTableColumn:[[NSTableColumn alloc] initWithIdentifier:@"member"]];

	[self.memberList awakeFromNib];

	self.controller = [IRCChannelMemberListController new];

	[self.controller setValue:self.memberList forKey:@"tableView"];

	[self.memberList setValue:self.controller forKey:@"contentController"];

	[self.controller replaceContents:@[]];
}

- (IRCChannelUser *)memberNamed:(NSString *)nickname modes:(NSString *)modes
{
	IRCUser *user = [[IRCUser alloc] initWithNickname:nickname onClient:self.client];

	IRCChannelUserMutable *member = [[IRCChannelUserMutable alloc] initWithUser:user];

	member.modes = modes;

	return [member copy];
}

- (void)insertMember:(IRCChannelUser *)member atIndex:(NSUInteger)index
{
	[self.controller insertObject:member atArrangedObjectIndex:index];
}

- (NSArray<NSString *> *)rowDescriptions
{
	NSMutableArray<NSString *> *rows = [NSMutableArray array];

	for (NSInteger row = 0; row < self.memberList.numberOfRows; row++) {
		if ([self.memberList isGroupRow:row]) {
			TVCMemberListSection *section = [self.memberList.dataSource tableView:self.memberList
														objectValueForTableColumn:nil
																			  row:row];

			[rows addObject:[NSString stringWithFormat:@"[%@]", section.title]];
		} else {
			IRCChannelUser *member = [self.memberList itemAtRow:row];

			[rows addObject:member.user.nickname];
		}
	}

	return rows;
}

- (void)testSingleRankIsAFlatList
{
	[self insertMember:[self memberNamed:@"alice" modes:@""] atIndex:0];
	[self insertMember:[self memberNamed:@"bob" modes:@""] atIndex:1];

	XCTAssertEqual(self.memberList.numberOfRows, 2);
	XCTAssertFalse([self.memberList isGroupRow:0]);
	XCTAssertEqualObjects([self rowDescriptions], (@[ @"alice", @"bob" ]));
}

- (void)testSecondRankAddsHeadersForEverySection
{
	[self insertMember:[self memberNamed:@"alice" modes:@""] atIndex:0];
	[self insertMember:[self memberNamed:@"bob" modes:@""] atIndex:1];

	/* The operator sorts first. */
	[self insertMember:[self memberNamed:@"carol" modes:@"o"] atIndex:0];

	XCTAssertEqualObjects([self rowDescriptions], (@[ @"[Operators]", @"carol", @"[Members]", @"alice", @"bob" ]));

	XCTAssertTrue([self.memberList isGroupRow:0]);
	XCTAssertTrue([self.memberList isGroupRow:2]);
	XCTAssertEqual([self.memberList rowForMemberAtIndex:0], 1);
	XCTAssertEqual([self.memberList rowForMemberAtIndex:1], 3);
	XCTAssertEqual([self.memberList rowForMemberAtIndex:2], 4);
	XCTAssertNil([self.memberList itemAtRow:2]);
}

- (void)testRowForItemMatchesItemAtRow
{
	IRCChannelUser *alice = [self memberNamed:@"alice" modes:@""];
	IRCChannelUser *carol = [self memberNamed:@"carol" modes:@"o"];
	IRCChannelUser *dave = [self memberNamed:@"dave" modes:@"v"];

	[self insertMember:alice atIndex:0];
	[self insertMember:carol atIndex:0];
	[self insertMember:dave atIndex:1];

	for (IRCChannelUser *member in @[ alice, carol, dave ]) {
		NSInteger row = [self.memberList rowForItem:member];

		XCTAssertEqual([self.memberList itemAtRow:row], member);
	}
}

- (void)testRemovingLastMemberOfASectionDropsItsHeader
{
	[self insertMember:[self memberNamed:@"alice" modes:@""] atIndex:0];
	[self insertMember:[self memberNamed:@"carol" modes:@"o"] atIndex:0];
	[self insertMember:[self memberNamed:@"dave" modes:@"v"] atIndex:1];

	XCTAssertEqualObjects([self rowDescriptions],
						  (@[ @"[Operators]", @"carol", @"[Voiced]", @"dave", @"[Members]", @"alice" ]));

	[self.controller removeObjectAtArrangedObjectIndex:1]; // dave

	XCTAssertEqualObjects([self rowDescriptions], (@[ @"[Operators]", @"carol", @"[Members]", @"alice" ]));

	[self.controller removeObjectAtArrangedObjectIndex:0]; // carol

	XCTAssertEqualObjects([self rowDescriptions], (@[ @"alice" ]));
	XCTAssertEqual(self.memberList.numberOfRows, 1);
}

- (void)testReplacingContentsRebuildsSections
{
	NSArray *members = @[
		[self memberNamed:@"carol" modes:@"o"],
		[self memberNamed:@"alice" modes:@""],
		[self memberNamed:@"bob" modes:@""],
	];

	[self.controller replaceContents:members];

	XCTAssertEqualObjects([self rowDescriptions], (@[ @"[Operators]", @"carol", @"[Members]", @"alice", @"bob" ]));

	[self.controller replaceContents:@[]];

	XCTAssertEqual(self.memberList.numberOfRows, 0);
}

- (void)testHeaderRowsAreNeitherSelectableNorTypeSelectable
{
	[self insertMember:[self memberNamed:@"alice" modes:@""] atIndex:0];
	[self insertMember:[self memberNamed:@"carol" modes:@"o"] atIndex:0];

	XCTAssertFalse([self.memberList.delegate tableView:self.memberList shouldSelectRow:0]);
	XCTAssertTrue([self.memberList.delegate tableView:self.memberList shouldSelectRow:1]);

	XCTAssertNil([self.memberList.delegate tableView:self.memberList typeSelectStringForTableColumn:nil row:0]);
	XCTAssertEqualObjects([self.memberList.delegate tableView:self.memberList typeSelectStringForTableColumn:nil row:1],
						  @"carol");
}

@end

NS_ASSUME_NONNULL_END

File: Tests/GlasstualTests/TVCMemberListSectionTests.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2008 - 2010 Satoshi Nakagawa <psychs AT limechat DOT net>
 * Copyright (c) 2010 - 2018 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

import AppKit
@testable import Glasstual
import Testing

@MainActor
@Suite("Member list sections")
struct TVCMemberListSectionTests {
	private let client: GLTTestClient
	private let memberList: MemberList
	private let controller: IRCChannelMemberListController

	init() {
		let client = GLTTestClient()
		let memberList = MemberList(frame: NSRect(x: 0, y: 0, width: 150, height: 400))
		memberList.addTableColumn(NSTableColumn(identifier: NSUserInterfaceItemIdentifier("member")))
		memberList.configure()

		let controller = memberList.contentController!
		controller.replaceContents([])

		self.client = client
		self.memberList = memberList
		self.controller = controller
	}

	@Test("Members of a single rank are shown as a flat list")
	func singleRankIsAFlatList() {
		insert(makeMember(named: "alice"), at: 0)
		insert(makeMember(named: "bob"), at: 1)

		#expect(memberList.numberOfRows == 2)
		#expect(memberList.isGroupRow(0) == false)
		#expect(rowDescriptions == ["alice", "bob"])
	}

	@Test("A second rank gives every section a header row")
	func secondRankAddsHeadersForEverySection() {
		insert(makeMember(named: "alice"), at: 0)
		insert(makeMember(named: "bob"), at: 1)
		insert(makeMember(named: "carol", modes: "o"), at: 0)

		#expect(rowDescriptions == ["[Operators]", "carol", "[Members]", "alice", "bob"])
		#expect(memberList.isGroupRow(0))
		#expect(memberList.isGroupRow(2))
		#expect(memberList.rowForMember(at: 0) == 1)
		#expect(memberList.rowForMember(at: 1) == 3)
		#expect(memberList.rowForMember(at: 2) == 4)
		#expect(memberList.item(atRow: 2) == nil)
	}

	@Test("A member's row round trips back to the member")
	func rowForItemMatchesItemAtRow() {
		let alice = makeMember(named: "alice")
		let carol = makeMember(named: "carol", modes: "o")
		let dave = makeMember(named: "dave", modes: "v")

		insert(alice, at: 0)
		insert(carol, at: 0)
		insert(dave, at: 1)

		for member in [alice, carol, dave] {
			let row = memberList.row(forItem: member)
			let item = memberList.item(atRow: row) as? ChannelUser

			#expect(item == member)
		}
	}

	@Test("Removing the last member of a section drops its header")
	func removingLastMemberOfASectionDropsItsHeader() {
		insert(makeMember(named: "alice"), at: 0)
		insert(makeMember(named: "carol", modes: "o"), at: 0)
		insert(makeMember(named: "dave", modes: "v"), at: 1)

		#expect(rowDescriptions == [
			"[Operators]", "carol", "[Voiced]", "dave", "[Members]", "alice",
		])

		controller.remove(atArrangedObjectIndex: 1)

		#expect(rowDescriptions == ["[Operators]", "carol", "[Members]", "alice"])

		controller.remove(atArrangedObjectIndex: 0)

		#expect(rowDescriptions == ["alice"])
		#expect(memberList.numberOfRows == 1)
	}

	@Test("Removing the only member leaves an empty flat list")
	func removingOnlyMemberLeavesAnEmptyFlatList() {
		insert(makeMember(named: "alice"), at: 0)

		controller.remove(atArrangedObjectIndex: 0)

		#expect(memberList.numberOfRows == 0)
		#expect(rowDescriptions.isEmpty)
	}

	@Test("Replacing the contents rebuilds every section")
	func replacingContentsRebuildsSections() {
		controller.replaceContents([
			makeMember(named: "carol", modes: "o"),
			makeMember(named: "alice"),
			makeMember(named: "bob"),
		])

		#expect(rowDescriptions == ["[Operators]", "carol", "[Members]", "alice", "bob"])

		controller.replaceContents([])

		#expect(memberList.numberOfRows == 0)
	}

	@Test("A header row can be neither selected nor typed to")
	func headerRowsAreNeitherSelectableNorTypeSelectable() throws {
		insert(makeMember(named: "alice"), at: 0)
		insert(makeMember(named: "carol", modes: "o"), at: 0)

		let delegate = try #require(memberList.delegate)

		#expect((delegate.tableView?(memberList, shouldSelectRow: 0) ?? true) == false)
		#expect(delegate.tableView?(memberList, shouldSelectRow: 1) ?? false)
		#expect(delegate.tableView?(memberList, typeSelectStringFor: nil, row: 0) == nil)
		#expect(delegate.tableView?(memberList, typeSelectStringFor: nil, row: 1) == "carol")
	}

	// MARK: - Snapshot diffs

	@Test("A join keeps the order and the person who was selected")
	func joinKeepsOrderAndSelection() {
		let alice = makeMember(named: "alice")
		let bob = makeMember(named: "bob")

		controller.replaceContents([alice, bob])
		select(bob)

		controller.insert(makeMember(named: "aaron"), atArrangedObjectIndex: 0)

		#expect(rowDescriptions == ["aaron", "alice", "bob"])
		#expect(selectedNicknames == ["bob"])
	}

	@Test("A part keeps the order and the person who was selected")
	func partKeepsOrderAndSelection() {
		let alice = makeMember(named: "alice")
		let bob = makeMember(named: "bob")
		let carol = makeMember(named: "carol")

		controller.replaceContents([alice, bob, carol])
		select(carol)

		controller.remove(atArrangedObjectIndex: 0)

		#expect(rowDescriptions == ["bob", "carol"])
		#expect(selectedNicknames == ["carol"])
	}

	@Test("A part takes the selection with it when the person selected left")
	func partClearsSelectionOfDepartedMember() {
		let alice = makeMember(named: "alice")
		let bob = makeMember(named: "bob")

		controller.replaceContents([alice, bob])
		select(alice)

		controller.remove(atArrangedObjectIndex: 0)

		#expect(rowDescriptions == ["bob"])
		#expect(selectedNicknames.isEmpty)
	}

	@Test("A rename keeps the row and the selection, because the person is the same")
	func renameKeepsRowAndSelection() {
		let alice = makeMember(named: "alice")
		let bob = makeMember(named: "bob")

		controller.replaceContents([alice, bob])
		select(alice)

		var renamed = alice
		renamed.user.nickname = "alicia"
		controller.replace(renamed, atArrangedObjectIndex: 0)

		#expect(rowDescriptions == ["alicia", "bob"])
		#expect(selectedNicknames == ["alicia"])
	}

	@Test("A mode change moves the member into its new section and the selection follows")
	func modeChangeMovesMemberAndKeepsSelection() {
		let alice = makeMember(named: "alice")
		let bob = makeMember(named: "bob")

		controller.replaceContents([alice, bob])
		select(bob)

		/* An op promotion resorts the member list, which reaches the table as one
		 new ordering. */
		var promoted = bob
		promoted.modes = "o"
		controller.replaceContents([promoted, alice])

		#expect(rowDescriptions == ["[Operators]", "bob", "[Members]", "alice"])
		#expect(selectedNicknames == ["bob"])
	}

	@Test("An edit in place redraws the row without moving it")
	func inPlaceEditKeepsTheRow() {
		let alice = makeMember(named: "alice")
		let bob = makeMember(named: "bob")

		controller.replaceContents([alice, bob])
		select(bob)

		var away = alice
		away.user.isAway = true
		controller.replace(away, atArrangedObjectIndex: 0)

		#expect(rowDescriptions == ["alice", "bob"])
		#expect(selectedNicknames == ["bob"])
		#expect((memberList.item(atRow: 0) as? ChannelUser)?.user.isAway == true)
	}

	@Test("A nickname is drawn on one line and truncated at the tail")
	func memberCellUsesSingleLineTailTruncation() throws {
		let cell = MemberListCell(frame: NSRect(x: 0, y: 0, width: 150, height: 28))
		let nicknameField = try #require(cell.textField)
		cell.configure()

		#expect(nicknameField.usesSingleLineMode)
		#expect(nicknameField.maximumNumberOfLines == 1)
		#expect(nicknameField.lineBreakMode == .byTruncatingTail)
	}

	@Test("An unused status image collapses to no width at all")
	func memberCellCollapsesUnusedStatusImage() throws {
		let cell = MemberListCell(frame: NSRect(x: 0, y: 0, width: 150, height: 28))
		let statusImageView = try #require(cell.subviews.compactMap { $0 as? NSImageView }.last)
		let statusWidthConstraint = try #require(
			statusImageView.constraints.first { $0.firstAttribute == .width }
		)

		cell.setStatusImageVisible(false)

		#expect(statusImageView.isHidden)
		#expect(statusWidthConstraint.constant == 0)

		cell.setStatusImageVisible(true)

		#expect(statusImageView.isHidden == false)
		#expect(statusWidthConstraint.constant == 16)
	}

	private var rowDescriptions: [String] {
		(0 ..< memberList.numberOfRows).compactMap { row in
			if let section = memberList.section(atRow: row) {
				return "[\(section.title)]"
			}

			return (memberList.item(atRow: row) as? ChannelUser)?.user.nickname
		}
	}

	private func makeMember(named nickname: String, modes: ChannelModeSymbolSet = "") -> ChannelUser {
		let user = User(nickname: nickname)
		var member = ChannelUser(user: user, prefixes: client.currentUserPrefixes)
		member.modes = modes

		return member
	}

	private func insert(_ member: ChannelUser, at index: Int) {
		controller.insert(member, atArrangedObjectIndex: index)
	}

	private var selectedNicknames: [String] {
		memberList.selectedRowIndexes.compactMap {
			(memberList.item(atRow: $0) as? ChannelUser)?.user.nickname
		}
	}

	private func select(_ member: ChannelUser) {
		let row = memberList.row(forItem: member)
		memberList.selectRowIndexes(IndexSet(integer: row), byExtendingSelection: false)
	}
}
